mlx/backend/metal/kernels/indexing.metal

```diff
@@ -36,29 +36,38 @@ inline size_t offset_neg_idx(uint32_t idx, size_t) {
   return idx;
 }
 
-template <typename T, typename IdxT, int NIDX>
+// IDX_NDIM is the number of dimensions of the indices arrays. Compile-time
+// special case for 0 and 1. Anything >= 2 uses the general case
+template <typename T, typename IdxT, int NIDX, int IDX_NDIM>
 [[kernel]] void gather(
     const device T *src [[buffer(0)]],
-    const device Indices<IdxT, NIDX>& indices [[buffer(1)]],
+    const constant Indices<IdxT, NIDX>& indices [[buffer(1)]],
     device T *out [[buffer(2)]],
-    const device int *src_shape [[buffer(3)]],
-    const device size_t *src_strides [[buffer(4)]],
-    const device size_t& src_ndim [[buffer(5)]],
-    const device int *slice_sizes [[buffer(6)]],
-    const device size_t& slice_size [[buffer(7)]],
-    const device int *axes [[buffer(8)]],
-    uint gid [[thread_position_in_grid]]) {
+    const constant int *src_shape [[buffer(3)]],
+    const constant size_t *src_strides [[buffer(4)]],
+    const constant size_t& src_ndim [[buffer(5)]],
+    const constant int *slice_sizes [[buffer(6)]],
+    const constant int *axes [[buffer(7)]],
+    uint2 index [[thread_position_in_grid]],
+    uint2 grid_dim [[threads_per_grid]]) {
 
-  auto ind_idx = gid / slice_size;
```

Comment:
It would be interesting to know whether the change in the access pattern (`ind_idx` now becomes the fast moving idx) also gives a speedup or simply the removal of the call to `elem_to_loc`.

Reply:
They are both critical actually but the access pattern is slightly more important. I didn’t benchmark the individual changes but it was something like 2x for just removing elemtoloc and another 5x from the access pattern